File: examples/canonical-auth/java/src/main/java/dev/tenantcore/auth/routes/MeController.java
package dev.tenantcore.auth.routes;

import dev.tenantcore.auth.auth.AuthContext;
import dev.tenantcore.auth.domain.InMemoryStore;
import dev.tenantcore.auth.domain.User;
import dev.tenantcore.auth.registry.RouteMetadata;
import dev.tenantcore.auth.registry.RouteRegistry;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.format.DateTimeFormatter;
import java.util.Map;

@RestController
public class MeController {

    private final InMemoryStore store;

    public MeController(InMemoryStore store) {
        this.store = store;
    }

    @GetMapping("/me")
    public ResponseEntity<Map<String, Object>> me() {
        AuthContext auth = (AuthContext) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        User user = store.getUserById(auth.sub());

        String tenantName = null;
        if (auth.tenantId() != null) {
            tenantName = store.getTenantName(auth.tenantId());
        }

        var allowedRoutesMapped = RouteRegistry.getAllowedRoutes(auth).stream()
            .map(this::toRoutePayload)
            .toList();

        Map<String, Object> responseMap = new java.util.LinkedHashMap<>();
        responseMap.put("sub", auth.sub());
        responseMap.put("email", auth.email());
        responseMap.put("display_name", user.displayName());
        responseMap.put("tenant_id", auth.tenantId() != null ? auth.tenantId() : null);
        responseMap.put("tenant_name", tenantName != null ? tenantName : null);
        responseMap.put("tenant_role", auth.tenantRole() != null ? auth.tenantRole() : null);
        responseMap.put("groups", auth.groups());
        responseMap.put("permissions", auth.permissions());
        responseMap.put("acl_ver", auth.aclVer());
        responseMap.put("allowed_routes", allowedRoutesMapped);
        responseMap.put("guide_sections", RouteRegistry.getGuideSections(auth));
        responseMap.put("issued_at", DateTimeFormatter.ISO_INSTANT.format(auth.issuedAt()));
        responseMap.put("expires_at", DateTimeFormatter.ISO_INSTANT.format(auth.expiresAt()));

        return ResponseEntity.ok(responseMap);
    }

    private Map<String, Object> toRoutePayload(RouteMetadata route) {
        return Map.of(
            "method", route.method(),
            "path", route.path(),
            "permission", route.permission() == null ? "" : route.permission(),
            "description", route.description(),
            "service", route.service(),
            "resource", route.resource(),
            "action", route.action()
        );
    }
}

File: examples/canonical-auth/elixir/lib/tenantcore_auth/controllers/me_controller.ex
defmodule TenantcoreAuth.MeController do
  use TenantcoreAuth, :controller

  alias TenantcoreAuth.Domain.InMemoryStore
  alias TenantcoreAuth.Registry.RouteRegistry

  def show(%Plug.Conn{assigns: %{auth: auth}} = conn, _params) do
    user = InMemoryStore.get_user_by_id(auth.sub)
    tenant_id =
      case auth.tenant_id do
        value when value in [nil, "null", :null] -> nil
        ~c"null" -> nil
        value -> value
      end

    json(conn, %{
      sub: auth.sub,
      email: auth.email,
      display_name: user && user.display_name,
      tenant_id: tenant_id,
      tenant_name: tenant_id && InMemoryStore.get_tenant_name(tenant_id),
      tenant_role: auth.tenant_role,
      groups: auth.groups,
      permissions: auth.permissions,
      acl_ver: auth.acl_ver,
      allowed_routes:
        RouteRegistry.get_allowed_routes(auth)
        |> Enum.map(&Map.take(&1, [:method, :path, :permission, :description])),
      guide_sections: RouteRegistry.get_guide_sections(auth),
      issued_at: DateTime.to_iso8601(auth.issued_at),
      expires_at: DateTime.to_iso8601(auth.expires_at)
    })
  end

  def show(conn, _params) do
    conn |> put_status(:unauthorized) |> json(%{error: "Unauthorized"})
  end
end
